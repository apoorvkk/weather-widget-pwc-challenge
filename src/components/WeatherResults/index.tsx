/*
This component is responsible for rendering the weather data under the form controls.
For example, it will render the current day's weather data (temp, humidity, pressure etc.)
and the forecasted weather for the upcoming week.
*/

import React from "react";
import dayjs from "dayjs";

import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

import { useWeatherConfig } from "../WeatherConfig/context";
import {
  ForecastWeekContainer,
  ForecastDayContainer,
  WeatherIconContainer,
} from "./styles";

const getWindDirection = (degree: number): string => {
  const directions = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"];

  return directions[Math.floor(degree / 45)];
};

const convertToFahrenheit = (c: number): number => (9 * c + 32 * 5) / 5;

const WeatherResults = (): React.ReactElement => {
  const {
    weatherData,
    isLoading,
    weatherDataStatus,
    isImperial,
  } = useWeatherConfig();

  if (weatherDataStatus !== 200 || isLoading) {
    return null;
  }

  const headline = weatherData.description.headline;
  const details = weatherData.description.details;
  const icon = weatherData.description.icon;
  const visibility = weatherData.visibility;
  const humidity = weatherData.humidity;
  const pressure = weatherData.pressure;
  const cloudiness = weatherData.cloudiness;
  const windSpeed = !isImperial
    ? weatherData.wind.speed
    : weatherData.wind.speed * 0.6214;
  const formattedWindSpeed = Number(windSpeed).toFixed(2);
  const windDirection = getWindDirection(weatherData.wind.degree);
  const windUnits = !isImperial ? "kph" : "mph";
  const currentTemp = !isImperial
    ? weatherData.currentTemp
    : convertToFahrenheit(weatherData.currentTemp);
  const tempUnits = !isImperial ? "C" : "F";

  return (
    <>
      <Row>
        <Col>
          <div>{dayjs().format("dddd, MMMM Do")}</div>
          <div>{headline}</div>
          <div>{details.charAt(0).toUpperCase() + details.slice(1)}</div>
        </Col>
      </Row>

      <Row>
        <Col>
          <Row>
            <WeatherIconContainer>
              <img
                alt="A picture of what the current weather is like."
                src={`http://openweathermap.org/img/wn/${icon}@2x.png`}
              />
            </WeatherIconContainer>
            <Col className="align-self-center">
              <h1>
                {Math.round(currentTemp)}&#176;{tempUnits}
              </h1>
            </Col>
          </Row>
        </Col>
        <Col>
          <ul>
            <li>{`Visibility: ${visibility} metres`}</li>
            <li>{`Humidity: ${humidity}%`}</li>
            <li>{`Wind: ${formattedWindSpeed} ${windUnits} ${windDirection}`}</li>
            <li>{`Pressure: ${pressure} hPa`}</li>
            <li>{`Cloudiness: ${cloudiness}%`}</li>
          </ul>
        </Col>
      </Row>

      <ForecastWeekContainer className="justify-content-center">
        {weatherData.days.slice(0, 7).map((dayData, i) => {
          let day = dayjs().add(i, "day").format("dddd");
          if (day === dayjs().format("dddd")) {
            day = "Today";
          }
          const minTemp = !isImperial
            ? dayData.min
            : convertToFahrenheit(dayData.min);
          const maxTemp = !isImperial
            ? dayData.max
            : convertToFahrenheit(dayData.max);
          const humidity = dayData.humidity;

          return (
            <ForecastDayContainer key={i}>
              <Row className="justify-content-center">{day}</Row>

              <Row className="justify-content-center">
                <img
                  src={`http://openweathermap.org/img/wn/${dayData.icon}@2x.png`}
                  alt={`A picture of what the forecasted weather is like on ${day}.`}
                />
              </Row>

              <Row className="justify-content-center">
                <span>{Math.round(maxTemp)}&#176;&nbsp;</span>
                <span className="text-muted">{Math.round(minTemp)}&#176;</span>
              </Row>
              <Row className="justify-content-center">
                <small className="text-muted">Humidity {humidity}%</small>
              </Row>
            </ForecastDayContainer>
          );
        })}
      </ForecastWeekContainer>
    </>
  );
};

export default WeatherResults;
