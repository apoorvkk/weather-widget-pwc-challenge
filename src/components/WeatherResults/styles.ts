import styled from "styled-components";

import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

const WeatherIconContainer = styled(Col)`
  flex-grow: 0;
`;

const ForecastDayContainer = styled(Col)`
  flex-grow: 0;
  margin-left: 10px;
  margin-right: 10px;
`;

const ForecastWeekContainer = styled(Row)`
  padding-bottom: 20px;
  padding-top: 20px;
  padding-left: 10px;
  padding-right: 10px;
  border-top: 1px solid #0000001f;
`;

export { ForecastWeekContainer, ForecastDayContainer, WeatherIconContainer };
