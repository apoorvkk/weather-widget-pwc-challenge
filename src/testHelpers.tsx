import React from "react";

import { render, RenderResult } from "@testing-library/react";
import _ from "lodash";
import axios from "axios";

import { weatherDataJson, forecastDataJson } from "./testData";
import { WeatherConfigContext } from "./components/WeatherConfig/context";

const defaultWeatherData = {
  days: [
    {
      humidity: 10,
      max: 40,
      min: 0,
      icon: "10e",
    },
    {
      humidity: 0,
      max: 60,
      min: 34,
      icon: "ee45",
    },
  ],
  currentTemp: 23.7,
  pressure: 10034,
  humidity: 89,
  cloudiness: 25,
  visibility: 1000,
  description: {
    headline: "Cloudy",
    details: "35% chain of rain",
    icon: "10d",
  },
  wind: {
    degree: 315,
    speed: 23,
  },
};
export const renderWithWeatherConfig = (
  ui: React.ReactElement,
  configOverrides = {}
): RenderResult => {
  const config = _.merge(
    {
      isLoading: false,
      isImperial: false,
      location: "",
      weatherData: defaultWeatherData,
      weatherDataStatus: 200,
      setIsImperial: () => {},
      setLocation: () => {},
      refreshWidget: async () => {},
    },
    configOverrides
  );

  return render(
    <WeatherConfigContext.Provider value={config}>
      {ui}
    </WeatherConfigContext.Provider>
  );
};

export const mockApiRequests = (
  weatherData = weatherDataJson,
  forecastData = forecastDataJson
): void => {
  (axios.get as jest.Mock) = jest.fn((url) => {
    if (url === "https://api.openweathermap.org/data/2.5/weather") {
      return Promise.resolve({
        status: 200,
        data: weatherData,
      });
    } else if (url === "https://api.openweathermap.org/data/2.5/onecall") {
      return Promise.resolve({
        status: 200,
        data: forecastData,
      });
    }
  });
};
