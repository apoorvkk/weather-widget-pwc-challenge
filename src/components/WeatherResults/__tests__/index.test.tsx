import React from "react";

import dayjs from "dayjs";

import { renderWithWeatherConfig } from "../../../testHelpers";

import WeatherResults from "../";

describe("when status is 404", () => {
  it("does not render any content", () => {
    const { container } = renderWithWeatherConfig(<WeatherResults />, {
      weatherDataStatus: 404,
    });

    expect(container).toMatchSnapshot();
  });
});

describe("when status is 500", () => {
  it("does not render any content", () => {
    const { container } = renderWithWeatherConfig(<WeatherResults />, {
      weatherDataStatus: 500,
    });

    expect(container).toMatchSnapshot();
  });
});

describe("when loading", () => {
  it("does not render any content", () => {
    const { container } = renderWithWeatherConfig(<WeatherResults />, {
      isLoading: true,
    });

    expect(container).toMatchSnapshot();
  });
});

describe("when results loaded successfully", () => {
  const testContent = (isImperial = false) => {
    const { queryByText, getByAltText } = renderWithWeatherConfig(
      <WeatherResults />,
      {
        isImperial,
      }
    );
    const nextDay = dayjs().add(1, "day").format("dddd");

    if (!isImperial) {
      expect(queryByText("24°C")).toBeInTheDocument();
      expect(queryByText("Wind: 23.00 kph NW")).toBeInTheDocument();
      expect(queryByText("40°")).toBeInTheDocument();
      expect(queryByText("0°")).toBeInTheDocument();
      expect(queryByText("60°")).toBeInTheDocument();
      expect(queryByText("34°")).toBeInTheDocument();
    } else {
      expect(queryByText("75°F")).toBeInTheDocument();
      expect(queryByText("Wind: 14.29 mph NW")).toBeInTheDocument();
      expect(queryByText("104°")).toBeInTheDocument();
      expect(queryByText("32°")).toBeInTheDocument();
      expect(queryByText("140°")).toBeInTheDocument();
      expect(queryByText("93°")).toBeInTheDocument();
    }

    expect(queryByText(dayjs().format("dddd, MMMM Do"))).toBeInTheDocument();
    expect(queryByText("Cloudy")).toBeInTheDocument();
    expect(queryByText("35% chain of rain")).toBeInTheDocument();
    expect(queryByText("Visibility: 1000 metres")).toBeInTheDocument();
    expect(queryByText("Humidity: 89%")).toBeInTheDocument();
    expect(queryByText("Pressure: 10034 hPa")).toBeInTheDocument();
    expect(queryByText("Cloudiness: 25%")).toBeInTheDocument();
    expect(
      getByAltText("A picture of what the current weather is like.")
    ).toHaveAttribute("src", "http://openweathermap.org/img/wn/10d@2x.png");
    expect(queryByText("Today")).toBeInTheDocument();
    expect(
      getByAltText("A picture of what the forecasted weather is like on Today.")
    ).toHaveAttribute("src", "http://openweathermap.org/img/wn/10e@2x.png");
    expect(queryByText("Humidity 10%")).toBeInTheDocument();
    expect(queryByText(nextDay)).toBeInTheDocument();
    expect(
      getByAltText(
        `A picture of what the forecasted weather is like on ${nextDay}.`
      )
    ).toHaveAttribute("src", "http://openweathermap.org/img/wn/ee45@2x.png");
    expect(queryByText("Humidity 0%")).toBeInTheDocument();
  };

  describe("when in metric format", () => {
    it("renders the results in metric format", () => {
      testContent(false);
    });
  });

  describe("when in imperial format", () => {
    it("renders the results in imperial format", () => {
      testContent(true);
    });
  });
});
