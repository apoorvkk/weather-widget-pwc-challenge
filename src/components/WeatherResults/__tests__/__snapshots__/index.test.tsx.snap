// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when loading does not render any content 1`] = `<div />`;

exports[`when status is 404 does not render any content 1`] = `<div />`;

exports[`when status is 500 does not render any content 1`] = `<div />`;
